### ISSUE TITLE
Check fastmath=True

### ISSUE THREAD

**Issue by seanlaw** [TDAmeritrade/stumpy#708]
In many `@njit` cases, we use `fastmath=True`. However, this might be bad when the input arguments and/or output results contain `np.nan` or `np.inf` (see [Numba fastmath](https://numba.pydata.org/numba-doc/latest/user/performance-tips.html#fastmath)). In those cases, we should use more specific but slower `fastmath` [flags](https://llvm.org/docs/LangRef.html#fast-math-flags) like `@njit(fastmath={"nsz", "arcp", "contract", "afn", "reassoc"})`

We should spend some time to reassessing whether all of the `fastmath` specifications are accurate or if they need to be updated

**Comment by seanlaw** [comment]
There is a nice article [here](https://simonbyrne.github.io/notes/fastmath/) on some of the differences in the fastmath flags. I wonder if it is possible/might be worth to go down a path where we assume that all `njit` functions must contain finite inputs/outputs. This will require some further preprocessing of the inputs AND post-processing of the outputs but it might be well worth the effort.

**Comment by seanlaw** [comment]
@NimaSarajpoor Did you actually spend time checking this in PR #1048 or does this work still need to happen?

**Comment by NimaSarajpoor** [comment]
@seanlaw 
I did not. So,  I think we should keep this issue open. 

IIRC, in #1025, we had some discussion around a code that finds caller(s) of njit function. We can add a function in `./fastmath.py` that raises an error if it finds a njit function whose fastmath flag is not a subset of the fastmath flag of its caller njit function(s). In addition, any njit function with `fastmath=True` needs to be checked [manually] to make sure it does not take an input that contains np.inf / np.nan.


What are your thought?

**Comment by seanlaw** [comment]
@NimaSarajpoor That sounds good.

> In addition, any njit function with fastmath=True needs to be checked [manually] to make sure it does not take an input that contains np.inf / np.nan.

Yes, this is probably the first step

> We can add a function in ./fastmath.py that raises an error if it finds a njit function whose fastmath flag is not a subset of the fastmath flag of its caller njit function(s).

Yes, adding a function to `fastmath.py` is fine. Is there anything I can assist/help you with? I think the current PR associated with this issue is too complicated and we should try to keep it simple/flat.

**Comment by NimaSarajpoor** [comment]
> I think the current PR associated with this issue is too complicated and we should try to keep it simple/flat.

I do agree. 

>  Is there anything I can assist/help you with?

Yes. It would be great if you could share with me what you think about the following steps:

(1) Close the old PR #1025. Create a new PR and add check list for all njit functions
(2) For each njit function `func`, check all of its callers to see if they pass an argument to `func` that may contain np.inf/np.nan 
(3) Have some discussion around adding the new function to `./fastmath`. I have some doubts whether this new function actually helps. We can discuss it further when we reach this point in the PR. 

I think this should keep the PR simple as you suggested.

**Comment by seanlaw** [comment]
> (1) Close the old PR https://github.com/TDAmeritrade/stumpy/pull/1025. Create a new PR and add check list for all njit functions

Sounds good.

> (2) For each njit function func, check all of its callers to see if they pass an argument to func that may contain np.inf/np.nan

I'm not sure what you mean by "check all of its callers to see if they pass an argument to func". Can you give me an example? Also, I'm assuming that this "checker" is NOT performing this check at runtime but, instead, at `test.sh` time?

> (3) Have some discussion around adding the new function to ./fastmath. I have some doubts whether this new function actually helps. We can discuss it further when we reach this point in the PR.

👍 Happy to talk about it whenever you think it is the right time.

**Comment by seanlaw** [comment]
@NimaSarajpoor I compiled the following list of njitted-functions that take `fastmath=config.STUMPY_FASTMATH_TRUE` and noted whether they should be switched:

- [ ] `aamp._compute_diagonal` - `P`, `PL`, `PR` contains `np.inf`  and `p` can be `np.inf`
- [ ] `aamp._aamp` - `P`, `PL`, `PR` contains `np.inf`
- [ ] `core._sliding_dot_product` - Should be okay
- [ ] `core._calculate_squared_distance_profile` - Should be okay
- [ ] `core._calculate_distance_profile` - Should be okay
- [ ] `core._p_norm_distance_profile` - Should be okay
- [ ] `core._mass` Should be okay
- [ ] `core._apply_exclusion_zone` - `val` contains `np.inf`
- [ ] `core._count_diagonal_ndist` - Should be okay
- [ ] `core._get_array_ranges` - Should be okay
- [ ] `core._get_ranges` - Should be okay
- [ ] `core._total_diagonal_ndists` - Should be okay
- [ ] `fastmath._add_assoc` - Should be okay
- [ ] `maamp._compute_multi_p_norm` - `p` could possibly be `np.inf`
- [ ] `mstump._compute_multi_D` - Might be okay??
- [ ] `scraamp._compute_PI` - `P_NORM` contains `np.inf`
- [ ] `scraamp._prescraamp` - `P_NORM` contains `np.inf`
- [ ] `scrump._compute_PI` - references `np.inf` values, so likely bad
- [ ] `scrump._prescrump` - `P_squared` is `np.inf`
- [ ] `stump._compute_diagonal` - `ρ`, `ρL`, and `ρR` contain `np.inf`
- [ ] `stump._stump` - `ρ`, `ρL`, and `ρR` contain `np.inf`

**Comment by NimaSarajpoor** [comment]
@seanlaw 
> I compiled the following list of njitted-functions that take fastmath=config.STUMPY_FASTMATH_TRUE and noted whether they should be switched:

That is very helpful! Thanks for going through them all!! I will go through those items. So, IIUC, the goal is to revise the fastmath flag for those that contains `n.p.inf`. Is that correct?

>  "check all of its callers to see if they pass an argument to func". Can you give me an example?

Suppose a njit function `func` has a param `x`. We want to know if `x` can contain `np.inf`. Let's say the function is something simple like:

```
@njit(fastmath=True)
def func(x):
     if x == np.inf:
         out = x
     else:
         out =  x + 1
     return out
 ```

Here I know that `x` can have `np.inf` and hence the fastmath flag needs to be revised. But, what if it is something like this:
```
@njit(fastmath=True)
def func(x):
     out =  x + 1
     return out
 ```
How do I know this is ok? I need to make sure the caller(s) of this function does not pass `x==inf` to this function. Based on the list you provided, I assume we do not have something like this in the STUMPY's code base and it is clear to recognize whether a param may contain `inf` or not just by looking at the function itself.


As a side:
What about the value that a callee function returns? For instance, the njit function `core.calculate_distance_profile` has the following lines:

https://github.com/TDAmeritrade/stumpy/blob/3676545d9d7622e7cbf79141b05f040d28a5954f/stumpy/core.py#L1226-L1230

If we follow the callee, we can see the array `D_squared` can contain `np.inf` value. Is that okay to do `np.sqrt(D_squared)` when `fastmath=True`?

---

> so, I'm assuming that this "checker" is NOT performing this check at runtime but, instead, at test.sh time?

I did not mean to write a "checker" but just manually check functions, and create a list, which is already provided in your recent comment!

**Comment by NimaSarajpoor** [comment]
> mstump._compute_multi_D - Might be okay??

This got my attention. I checked the function and noticed there are some computations that involve QT

https://github.com/TDAmeritrade/stumpy/blob/3676545d9d7622e7cbf79141b05f040d28a5954f/stumpy/mstump.py#L911-L917

And, from the function itself, it is hard to tell whether QT contains `np.inf` or not. So...

>  I need to make sure the caller(s) of this function does not pass x==inf to this function. 

So, if I follow the callers of this function, I can see that `QT` is computed here: 

https://github.com/TDAmeritrade/stumpy/blob/3676545d9d7622e7cbf79141b05f040d28a5954f/stumpy/mstump.py#L1264

and it cannot contain `np.inf` value.

**Comment by seanlaw** [comment]
> That is very helpful! Thanks for going through them all!! I will go through those items. So, IIUC, the goal is to revise the fastmath flag for those that contains np.inf. Is that correct?

Yes, according to the [LLVM fastmath flags](https://llvm.org/docs/LangRef.html#fast-math-flags), if `fastmath=True`:

> This flag is a shorthand for specifying all fast-math flags at once, and imparts no additional semantics from using all of them.

This includes `nnan` and `ninf`, which means that neither the arguments (inputs?) and results (outputs) are not `nan`/`+/- inf`. So, in the cases where I say "Should be okay", then I think they don't have any non-finite arguments/results and therefore we can probably use `fastmath=True`. Otherwise, all of the other ones must use `fastmath=config.STUMPY_FASTMATH_FLAGS` as they likely have non-finite arguments/results and are therefore unsafe to use `fastmath=True`.

**Comment by seanlaw** [comment]
> How do I know this is ok? I need to make sure the caller(s) of this function does not pass x==inf to this function. 

We don't know if it is okay just from examining the function itself. We must either, roughly, examine the logic of each caller/callee or, if we are lazy, we must concede and use `fastmath=config.STUMPY_FASTMATH_FLAGS` everywhere.

> Based on the list you provided, I assume we do not have something like this in the STUMPY's code base and it is clear to recognize whether a param may contain inf or not just by looking at the function itself.

So, my logic is this:

1. We manually check (i.e., this cannot be automated!!) whether the arguments/results (i.e., inputs/outputs) may possibly be non-finite so that we can label each function as `fastmath=config.STUMPY_FASTMATH_TRUE` or `fastmath=config.STUMPY_FASTMATH_FLAGS`
2. Note that a caller (parent) may pass non-finite values down to a callee (child) AND, conversely, a callee (child) may pass non-finite values up to a caller (parent). Therefore, we must always consider the non-finite "poison" being passed in EITHER/BOTH direction!
3. Because of this, we can take an entire call chain, collect all of the `fastmath` flags in that chain into one Python `set`, and count the number of "distinct"/"unique" `fastmath` flags. If they are all the same then `len(set) == 1`. If they are non-uniform then `len(set) >= 2` and this means that we should likely flip all of the flags within this chain to `fastmath=config.STUMPY_FASTMATH_FLAGS`
4. Since flipping flags may affect OTHER calling chains, we MUST repeat step 3 again and again until we no longer find `len(set) >= 2` (i.e., all calling chains are "pure" - `len(set) == 1`)

That's roughly what I am thinking. It's possible that EVERYTHING becomes `fastmath=config.STUMPY_FASTMATH_FLAGS`... However, this is probably not true as we have simple njit functions that aren't nested (i.e., they have no child callee and their parent caller is a non-njit function).

**Comment by seanlaw** [comment]
> And, from the function itself, it is hard to tell whether QT contains np.inf or not. So...

Yeah, this stuff is very tricky/impossible to check. It's better to be overly conservative as there is no easy way to check all of the variables coming in-and-out without putting a huge burden on ourselves

**Comment by NimaSarajpoor** [comment]
@seanlaw 

>  in the cases where I say "Should be okay", then I think they don't have any non-finite arguments/results and therefore we can probably use fastmath=True. Otherwise, all of the other ones must use fastmath=config.STUMPY_FASTMATH_FLAGS

👍 
 
>  How do I know this is ok? I need to make sure the caller(s) of this function does not pass x==inf to this function.
> > We don't know if it is okay just from examining the function itself. 

> Yeah, this stuff is very tricky/impossible to check. 

> Note that a caller (parent) may pass non-finite values down to a callee (child) AND, conversely, a callee (child) may pass non-finite values up to a caller (parent). Therefore, we must always consider the non-finite "poison" being passed in EITHER/BOTH direction!

Right on!


> Is there a flaw in my logic? It may not be perfect

When you said:  "we can take an entire call chain, collect all of the fastmath flags in that chain into one Python set, and count the number of distinct/unique fastmath flags. If they are all the same then `len(set) == 1`. If they are non-uniform then `len(set) >= 2` and this means that we should likely flip all of the flags within this chain to `fastmath=config.STUMPY_FASTMATH_FLAGS.`"

Once the set is obtained, then we need to check how the chain is. I think that is why you said "should likely flip" rather than "must flip". Because, it is possible that we pass argument to a njit callee but the actual computation is happening at a deeper level in another function which is called by the callee. Here we are having that "burden" you mentioned. So, I agree with:

> . It's better to be overly conservative

**Comment by NimaSarajpoor** [comment]
@seanlaw 
I believe the discussion has been very helpful! I am going to submit a new PR and will add [the list](https://github.com/TDAmeritrade/stumpy/issues/708#issuecomment-2613757160) you provided. Then, I can start making changes gradually. I will try to not deviate from the main track and keep the PR simple and clear as you suggested before.

**Comment by seanlaw** [comment]
Sounds good @NimaSarajpoor. If we find, after we check all of the call chains and make all of the changes, that 90% or more of the functions require `fastmath=config.STUMPY_FASTMATH_FLAGS` then we might as well just flip all of the `fastmath=config.STUMPY_FASTMATH_TRUE` to `fastmath=config.STUMPY_FASTMATH_FLAGS`. It would mean fewer assumptions/checks (at the cost of a little speed reduction) and result in easier long term maintenance without any "gotchas".
